### DISCUSSION TITLE
TextField onChange not working when using IdentifiedArrayOf

### DISCUSSION THREAD

**Discussion by cadizjavier** [pointfreeco/swift-composable-architecture#2668] [Q&A]
The `onChange` modifier on a `TextField` doesn't work when using `IdentifiedArrayOf`.

The minimal example I was able to create is a dynamic Form using the `BindingReducer`
Let's take a look at it. (Working on the `observation-beta` branch)

```swift
// ContentView.swift
import ComposableArchitecture
import SwiftUI

struct Account: Identifiable, Equatable {
    let id: UUID
    let name: String
    var balance: String
}

@Reducer
struct ContentFeature {
    @ObservableState
    struct State {
        var contacts: IdentifiedArrayOf<Account> = []
    }

    enum Action: BindableAction, Equatable {
        case binding(BindingAction<State>)
    }

    var body: some ReducerOf<Self> {
        BindingReducer()
        Reduce { state, action in
            switch action {
            case .binding:
                return .none
            }
        }
    }
}

struct ContentView: View {
    @Bindable var store: StoreOf<ContentFeature>

    var body: some View {
        Form {
            ForEach($store.contacts) { $contact in
                HStack {
                    Text(contact.name)
                    Spacer()
                    TextField(contact.name, text: $contact.balance)
                        .keyboardType(.numberPad)
                        .multilineTextAlignment(.trailing)
                        .onChange(of: contact.balance) { oldValue, newValue in
                            print(newValue)
                        }
                }
            }
        }
    }
}
```

```swift
// DemoApp.swift
import ComposableArchitecture
import SwiftUI

@main
struct DemoApp: App {
    var body: some Scene {
        WindowGroup {
            ContentView(store: Store(
                initialState: ContentFeature.State(
                    contacts: [
                        .init(id: UUID(), name: "Account 1", balance: "20"),
                        .init(id: UUID(), name: "Account 2", balance: "15")
                    ]
                )
            ) {
                ContentFeature()
            })
       }
    }
}
```

When the application runs and we modify a TextField value, the `onChange` modifier isn't triggered.

If we make this change:

```diff
- var contacts: IdentifiedArrayOf<Account> = []
+ var contacts: [Account] = []
```

It works. When changing TextField values, the `onChange` modifier gets triggered.

I'm unsure if this is as expected or is a limitation of `IdentifiedArrayOf.`
I hoped to use `IdentifiedArrayOf` for the benefits explained in the motivation plan.

Any hints, feedback, or better ways to do this would be appreciated.

**Comment by stephencelis** [comment]
@cadizjavier I copied and pasted the code above into a fresh project off the latest `observation-beta` and the `onChange` was firing just fine (Xcode 15.1, iOS 17.2 simulator), so I'm not quite sure why you're seeing the behavior you're seeing. Might need to clean your build or derived data or even restart? Or maybe it's related to a particular platform/OS.

**Reply by cadizjavier** [reply]
I just created a brand new project, and as you said. It works!!

I was curious about the discrepancies and noticed that I hadn't updated the package since I created the project when Observation Beta was announced.

Updating the `swift-composable-architecture` package fixed my issue.

For reference, If we use the package from the `bb063bd2bf86288b139a5b3987d6eaa2880f4e05` commit, which was more or less the date I created the project (27 Nov), the mentioned snippet doesn't work.

Note to myself. Be sure to stay on the latest commit before reporting a bug or asking a question. 😅

Thanks for the quick reply 🙏
